From 3ec27b7519e93edbdf2cd9d8a4d2252558ea3199 Mon Sep 17 00:00:00 2001
From: Nithya <nithya.mani@caviumnetworks.com>
Date: Wed, 10 Feb 2016 18:00:56 -0800
Subject: [PATCH 0382/1239] Read the board type from DT passed by BDK

Changed the u-boot implementation of reading the board
name from device tree passed by BDK instead of env_ptr.
Recently env_ptr has been changed in ATF and this
causes crash in u-boot.

Signed-off-by: Nithya <nithya.mani@caviumnetworks.com>
---
 board/cavium/thunderx/atf.c      | 10 +++++++
 board/cavium/thunderx/thunderx.c | 50 +++++++++++++++++++++++++++++++-
 drivers/net/cavium/thunder_bgx.c | 10 +++----
 include/cavium/atf.h             |  2 +-
 include/cavium/thunderx_svc.h    |  3 ++
 5 files changed, 68 insertions(+), 7 deletions(-)

diff --git a/board/cavium/thunderx/atf.c b/board/cavium/thunderx/atf.c
index 2e7ba69d76..4c47ba9a6f 100644
--- a/board/cavium/thunderx/atf.c
+++ b/board/cavium/thunderx/atf.c
@@ -155,6 +155,16 @@ ssize_t atf_env_string(size_t index, char *str)
 	}
 }
 
+ssize_t atf_get_bdk_fdt(void *buffer, size_t buffer_size)
+{
+	struct pt_regs regs;
+	regs.regs[0] = THUNDERX_FDT_GET;
+	regs.regs[1] = (uintptr_t)buffer;
+	regs.regs[2] = buffer_size;
+	smc_call(&regs);
+	return regs.regs[0];
+}
+
 #ifdef CONFIG_CMD_ATF
 
 static void atf_print_ver(void)
diff --git a/board/cavium/thunderx/thunderx.c b/board/cavium/thunderx/thunderx.c
index ab722ea3a3..3f88753aaa 100644
--- a/board/cavium/thunderx/thunderx.c
+++ b/board/cavium/thunderx/thunderx.c
@@ -59,6 +59,29 @@ DECLARE_GLOBAL_DATA_PTR;
 
 #define BOARD_TYPE "BOARD="
 
+#ifdef CONFIG_OF_LIBFDT
+static const void *get_prop_value(void *fdt, const char *prop_name, int *len)
+{
+	int depth = 0, node;
+	node = fdt_next_node(fdt, 0, &depth);
+
+	while (node >= 0) {
+		int prop_off;
+		prop_off = fdt_first_property_offset(fdt, node);
+		while (prop_off >= 0) {
+			const char *name;
+			const void *val = fdt_getprop_by_offset(fdt, prop_off,
+						&name, len);
+			if (strcmp(name, prop_name) == 0)
+				return val;
+			prop_off = fdt_next_property_offset(fdt, prop_off);
+		}
+		node = fdt_next_node(fdt, node, &depth);
+	}
+	return NULL;
+}
+#endif
+
 int board_init(void)
 {
 #ifdef CONFIG_OF_LIBFDT
@@ -114,6 +137,31 @@ void reset_cpu(ulong addr)
  */
 int board_late_init(void)
 {
+#ifdef CONFIG_OF_LIBFDT
+	char boardname[32];
+	const char *str;
+	void *fdt = (void *)CONFIG_BDK_FDT_START;
+	int ret = 0, len = 32;
+
+	atf_get_bdk_fdt(fdt, CONFIG_BDK_FDT_SIZE);
+
+	if (fdt != NULL) {
+		ret = fdt_check_header(fdt);
+		if (ret < 0) {
+			printf("fdt: %s\n", fdt_strerror(ret));
+		} else {
+			debug("fdt:size %d\n", fdt_totalsize(fdt));
+			str = get_prop_value(fdt, "BOARD-MODEL", &len);
+			debug("fdt:len %d\n", len);
+			if (str) {
+				strncpy(boardname, str, len);
+				setenv("board", boardname);
+			} else {
+				printf("Err: cannot retrieve board type from fdt\n");
+			}
+		}
+	}
+#else
 	int i;
 	char str[32];
 	const char *boardname;
@@ -129,7 +177,7 @@ int board_late_init(void)
 			break;
 		}
 	}
-
+#endif
 	printf("Board type: %s\n", getenv("board"));
 
 	return 0;
diff --git a/drivers/net/cavium/thunder_bgx.c b/drivers/net/cavium/thunder_bgx.c
index ce9d2539ed..ab19dec274 100644
--- a/drivers/net/cavium/thunder_bgx.c
+++ b/drivers/net/cavium/thunder_bgx.c
@@ -563,9 +563,9 @@ static int bgx_lmac_count(struct bgx *bgx)
 	const char *boardtype = getenv("board");
 	int lmac_count = 0;
 
-	if (!strcmp(boardtype, "ebb8800") ||
-	    !strcmp(boardtype, "ebb8804") ||
-	    !strcmp(boardtype, "ebb8604")) {
+	if (!strcasecmp(boardtype, "ebb8800") ||
+	    !strcasecmp(boardtype, "ebb8804") ||
+	    !strcasecmp(boardtype, "ebb8604")) {
 		switch (bgx->qlm_mode) {
 		case QLM_MODE_SGMII:
 		case QLM_MODE_XFI_4X1:
@@ -581,7 +581,7 @@ static int bgx_lmac_count(struct bgx *bgx)
 			lmac_count = 2;
 			break;
 		}
-	} else if (!strcmp(boardtype, "crb_1s")) {
+	} else if (!strcasecmp(boardtype, "crb_1s")) {
 		switch (bgx->bgx_id) {
 		case 0:
 			lmac_count = 1;
@@ -590,7 +590,7 @@ static int bgx_lmac_count(struct bgx *bgx)
 			lmac_count = 2;
 			break;
 		}
-	} else if (!strcmp(boardtype, "crb_2s")) {
+	} else if (!strcasecmp(boardtype, "crb_2s")) {
 		lmac_count = 1;
 	} else {
 		printf("Unsupported board\n");
diff --git a/include/cavium/atf.h b/include/cavium/atf.h
index 3cf05c43d7..1658e9c5a5 100644
--- a/include/cavium/atf.h
+++ b/include/cavium/atf.h
@@ -17,5 +17,5 @@ ssize_t atf_dram_size(unsigned int node);
 ssize_t atf_node_count(void);
 ssize_t atf_env_count(void);
 ssize_t atf_env_string(size_t index, char *str);
-
+ssize_t atf_get_bdk_fdt(void *buffer, size_t buffer_size);
 #endif
diff --git a/include/cavium/thunderx_svc.h b/include/cavium/thunderx_svc.h
index de36d66230..a75b968a13 100644
--- a/include/cavium/thunderx_svc.h
+++ b/include/cavium/thunderx_svc.h
@@ -63,4 +63,7 @@
 
 #define THUNDERX_NODE_COUNT		0x43000601
 
+/* x1 - user buffer, x2 - buffer size */
+#define THUNDERX_FDT_GET                0x43000701
+
 #endif /* __THUNDERX_SVC_H__ */
-- 
2.29.0

